### PR TITLE
Obtaining individual UTXOs from BlockFlowGroupView for SelectionAlgo

### PR DESCRIPTION
**Pull request by pragmaxim**
Currently we could retrieve only all UTXOs as `AssetOutputInfo`, let's make it possible even for specified UTXOs.

### CODE REVIEW

File: flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala [hunk 1]
```diff
@@ -141,18 +127,65 @@ object BlockFlowGroupView {
...
+                        outputRef,
+                        output,
+                        UnpersistedBlockOutput
+                      )
+                    )
+                  )
+                case Some(_) =>
+                  Left(WorldState.expectedAssetError)
+                case None =>
+                  Right(None)
```
Comment (polarker) [inline flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala:185]: @pragmaxim let's add a simple test to cover this branch. In fact, I am surprised this is not covered since we have so many tests.

File: flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala [hunk 2]
```diff
@@ -100,9 +100,6 @@ trait BlockFlowGroupView[WS <: WorldState[_, _, _, _]] {
       errorIfExceedMaxUtxos: Boolean
   ): IOResult[AVector[AssetOutputInfo]]
 
-  def getContractUtxos(lockupScript: LockupScript.P2C): IOResult[ContractOutput] = {
-    worldState.getContractAsset(lockupScript.contractId)
-  }
```
Comment (pragmaxim) [inline flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala:119]: Dead code elimination

File: flow/src/test/scala/org/alephium/flow/core/BlockFlowSpec.scala [hunk 3]
```diff
@@ -988,79 +988,14 @@ class BlockFlowSpec extends AlephiumSpec {
...
-  }
-
   it should "not store txOutputRef index when it is disabled" in new TxOutputRefIndexFixture {
     override def enableTxOutputRefIndex = false
+    testTxScriptCalling()
   }
 
   it should "store txOutputRef index when it is enabled" in new TxOutputRefIndexFixture {
     override def enableTxOutputRefIndex = true
+    testTxScriptCalling()
```
Comment (pragmaxim) [inline flow/src/test/scala/org/alephium/flow/core/BlockFlowSpec.scala:998]: Made this to let devs test the intermediate state before calling script

File: flow/src/test/scala/org/alephium/flow/core/BlockFlowSpec.scala [hunk 4]
```diff
@@ -1078,3 +1013,77 @@ class BlockFlowSpec extends AlephiumSpec {
     } is true
   }
 }
+
+trait TxOutputRefIndexFixture extends FlowFixture {
```
Comment (pragmaxim) [inline flow/src/test/scala/org/alephium/flow/core/BlockFlowSpec.scala:1017]: Moved Fixture out so it is accessible also from `BlockFlowGroupViewSpec`

File: flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala (hunk 1 above)
Comment (pragmaxim) [reply flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala:182]: This one cannot happen @polarker as if `blockCaches(index).relatedOutputs.get(outputRef)` returned `None`, the condition above `blockCaches.indexWhere(_.relatedOutputs.contains(outputRef))` would return `-1` and this branch would not be executed, also the array would get illegal index anyway.

Comment (pragmaxim) [reply flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala]: After this chance I am quite sure that codecov has high chance for making a false positive as it behaves non-deterministicaly.

**Comment by polarker** [comment]
@h0ngcha0 Please help review the new commits that improved the test coverage

**Review by h0ngcha0 — APPROVED** [review]
LGTM :+1:

File: flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala [hunk 5]
```diff
@@ -134,19 +134,15 @@ object BlockFlowGroupView {
...
-            if (index != -1) {
-              blockCaches(index).relatedOutputs.get(outputRef) match {
-                case Some(output: ContractOutput) =>
-                  Right(Some(output))
-                case Some(_) =>
-                  Left(WorldState.expectedContractError)
-                case None =>
-                  Right(None)
+            blockCaches.iterator
+              .collectFirst {
```
Comment (polarker) [inline flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala:138]: Let's revert this change. Scala std collection's higher-order functions are generally slow.

Since we know why that line won't be covered, it's fine that we leave it as it is.

Comment (polarker) [reply flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala:138]: Or we could introduce `collectFirst` for `AVector`

File: flow/src/test/scala/org/alephium/flow/core/BlockFlowGroupViewSpec.scala [hunk 6]
```diff
@@ -20,13 +20,21 @@ import scala.collection.mutable
...
+  GroupIndex,
+  Hint,
+  TxOutputRef
+}
+import org.alephium.util.{AVector, TimeStamp}
 
-class BlockFlowGroupViewSpec extends AlephiumSpec {
-  it should "fetch preOutputs" in new FlowFixture {
+class BlockFlowGroupViewSpec extends AlephiumFlowSpec {
+  it should "fetch getPreAssetOutputs and getPreOutputs" in new FlowFixture {
```
Comment (polarker) [inline flow/src/test/scala/org/alephium/flow/core/BlockFlowGroupViewSpec.scala:37]: It seems getPreContractOutput is tested instead of getPreOutputs?

File: flow/src/test/scala/org/alephium/flow/core/TxUtilsSpec.scala [hunk 7]
```diff
@@ -248,6 +249,62 @@ class TxUtilsSpec extends AlephiumSpec {
...
+      // it should be absent in mempool
+      val tx = block.nonCoinbase.head
+      blockFlow.getGrandPool().add(chainIndex, tx.toTemplate, TimeStamp.now())
+      blockFlow
+        .getMutableGroupViewIncludePool(fromGroup)
+        .rightValue
+        .getPreContractOutput(contractOutputRef)
+        .rightValue
+        .isEmpty is true
+
```
Comment (polarker) [inline flow/src/test/scala/org/alephium/flow/core/TxUtilsSpec.scala:291]: extra empty line

File: flow/src/test/scala/org/alephium/flow/core/BlockFlowGroupViewSpec.scala (hunk 6 above)
Comment (pragmaxim) [reply flow/src/test/scala/org/alephium/flow/core/BlockFlowGroupViewSpec.scala:37]: There were preOutputs tests already but only for `additionalCache`, adding the missing case + moving `getPreContractOutput` for missing contractRef to a better place, you are right :+1:

File: util/src/test/scala/org/alephium/util/AVectorSpec.scala [hunk 8]
```diff
@@ -511,6 +511,17 @@ class IntAVectorSpec extends AVectorSpec[Int] {
...
 
+  it should "collectFirst" in new FixtureF {
+    AVector.empty[Int].collectFirst { case i if i > 0 => i * i } shouldBe None
+
+    AVector(-1, 2, 3).collectFirst { case i if i > 0 => i * i } shouldBe Some(4)
+
+    AVector(-1, -2, -3).collectFirst { case i if i > 0 => i * i } shouldBe None
+
+    // Case where the first match is returned, even if there are other matches
+    AVector(1, 2, 3).collectFirst { case i if i > 1 => i * i } shouldBe Some(4)
```
Comment (polarker) [inline util/src/test/scala/org/alephium/util/AVectorSpec.scala]: Isn't this the same as test case 2?

File: flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala [hunk 9]
```diff
@@ -135,12 +135,10 @@ object BlockFlowGroupView {
             Left(WorldState.expectedContractError)
           case None =>
             blockCaches
-              .collectFirst {
-                case blockCache if blockCache.relatedOutputs.contains(outputRef) =>
-                  blockCache.relatedOutputs(outputRef) match {
```
[the remark this answers is not in the document]
Comment (pragmaxim) [reply flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala]: Can you clarify? Turning `PartionFunction` to `Option` in collect/collectFirst will probably always make everyone think why it is like this.

[the remark this answers is not in the document]
Comment (pragmaxim) [reply flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala]: Ou I see now, `isDefinedAt` is tricky and using `A => Option[B]` is probably cleaner.

File: flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala [hunk 10]
```diff
@@ -252,13 +252,6 @@ object BlockFlowGroupView {
       mempool: MemPool
   ) extends Impl0[WS](worldState, blockCaches) {
 
-    override def getPreContractOutput(
-        outputRef: ContractOutputRef
-    ): IOResult[Option[ContractOutput]] = {
-      // MemPool ignores ContractOutputs
-      super.getPreContractOutput(outputRef)
-    }
```
Comment (polarker) [inline flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala]: This is removed now.

Comment (pragmaxim) [reply flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala]: Agree :+1: for me it was 50/50, I decided to make it more explicit
